### PR TITLE
Auto generate node ID

### PR DESCRIPTION
**Pull request by honnix**
# TL;DR
Auto generate node ID

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
In most cases, there is no need to specify a node ID that must be unique across one workflow. This PR makes it possible to auto-generate one if unspecified.

If node ID is not specified, it will be generated as `w<random>-n0`, `w<random>-n1`, etc. where `random` is a random integer generated per workflow build, and `w<random>` can be specified by a user; and to make a node readable, node name is inferred from class name.

## Tracking Issue
Closes https://github.com/flyteorg/flyte/issues/3185

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkCondition.java [hunk 1]
```diff
@@ -54,13 +54,6 @@ public SdkBranchNode apply(
       List<String> upstreamNodeIds,
       @Nullable SdkNodeMetadata metadata,
       Map<String, SdkBindingData> inputs) {
-    if (metadata != null) {
```
Comment (honnix) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkCondition.java:57]: This feels unnecessarily strict.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkContainerTask.java [hunk 2]
```diff
@@ -53,11 +53,6 @@ public String getType() {
     return "raw-container";
   }
 
-  /** Specifies task name. */
-  public String getName() {
```
Comment (honnix) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkContainerTask.java:57]: Moved to `SdkTransform`.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkNode.java [hunk 3]
```diff
@@ -58,6 +58,6 @@ public SdkNode apply(String id, SdkTransform transform) {
     List<String> upstreamNodeIds =
         getOutputs().isEmpty() ? Collections.singletonList(getNodeId()) : Collections.emptyList();
 
-    return builder.applyInternal(id, transform, upstreamNodeIds, /*metadata=*/ null, getOutputs());
```
Comment (honnix) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkNode.java:61]: This is strictly a breaking change, but I'm not sure why we had `applyInternal` as `protected`, and here `metadata` is always `null`.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkPartialTransform.java [hunk 4]
```diff
@@ -114,6 +126,11 @@ public SdkTransform withTimeoutOverride(Duration timeout) {
     return new SdkPartialTransform(transform, fixedInputs, extraUpstreamNodeIds, mergedMetadata);
   }
 
+  @Override
+  public String getName() {
+    return transform.getName();
```
Comment (honnix) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkPartialTransform.java:131]: Delegate to `transform` to get real name.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java [hunk 5]
```diff
@@ -47,34 +54,52 @@ public SdkWorkflowBuilder() {
...
-      String nodeId,
+  public SdkNode apply(SdkTransform transform) {
+    return apply(/*nodeId=*/ null, transform, emptyMap());
+  }
+
+  public SdkNode apply(SdkTransform transform, Map<String, SdkBindingData> inputs) {
+    return applyInternal(/*nodeId=*/ null, transform, emptyList(), inputs);
+  }
+
+  SdkNode applyInternal(
```
Comment (honnix) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java:77]: Changed visibility to package only. Note that this is strictly a breaking change.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java [hunk 6]
```diff
@@ -201,4 +226,9 @@ public void output(String name, SdkBindingData value, String help) {
   public WorkflowTemplate toIdlTemplate() {
     return WorkflowTemplateIdl.ofBuilder(this);
   }
+
+  private static String toNodeName(String name) {
+    String lastPart = name.substring(name.lastIndexOf('.') + 1);
+    return PATTERN.matcher(lastPart).replaceAll("$1-$2").toLowerCase().replaceAll("\\$", "-");
```
Comment (honnix) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java:232]: Basically camel to kebab, also placing `$` for nested class.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java [hunk 7]
```diff
@@ -47,34 +55,52 @@ public SdkWorkflowBuilder() {
...
               CompilerError.Kind.DUPLICATE_NODE_ID,
-              nodeId,
+              actualNodeId,
               "Trying to insert two nodes with the same id.");
 
       throw new CompilerException(error);
     }
 
-    SdkNode sdkNode = transform.apply(this, nodeId, upstreamNodeIds, metadata, inputs);
+    String fallbackNodeName =
```
Comment (honnix) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java:97]: For backward compatibility.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkCondition.java (hunk 1 above)
Comment (honnix) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkCondition.java:57]: I had to delete these because of calling `SdkTransform withNameOverride(String name, boolean failOnDuplicate)`

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java [hunk 8]
```diff
@@ -26,55 +26,89 @@
...
   private final Map<String, SdkNode> nodes;
   private final Map<String, SdkBindingData> inputs;
   private final Map<String, SdkBindingData> outputs;
   private final Map<String, String> inputDescriptions;
   private final Map<String, String> outputDescriptions;
+  private final String nodeIdPrefix;
+  private final AtomicInteger nodeIdSuffix;
 
   public SdkWorkflowBuilder() {
+    this("w" + ThreadLocalRandom.current().nextInt(1000, 10000) + "-");
```
Comment (honnix) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java:53]: Make sure we could support subworkflow later.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java [hunk 9]
```diff
@@ -72,6 +72,10 @@ public SdkTransform withUpstreamNode(SdkNode node) {
   }
 
   public SdkTransform withNameOverride(String name) {
+    return withNameOverride(name, true);
+  }
+
+  SdkTransform withNameOverride(String name, boolean failOnDuplicate) {
```
Comment (honnix) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java:78]: Intentionally not opening this method to avoid exposing unnecessary methods to users.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java (hunk 8 above)
Comment (honnix) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java:53]: I went for consistency because we use number for node already `n0, `n1` etc. This is a also bit simpler than generating random bytes. But I'm ok either way.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java [hunk 10]
```diff
@@ -72,6 +72,10 @@ public SdkTransform withUpstreamNode(SdkNode node) {
   }
 
   public SdkTransform withNameOverride(String name) {
+    return withNameOverride(name, true);
+  }
```
Comment (ckiosidis) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java:76]: this is exposed publicly? So now we have 2 ways to set the name? One though the withNameOverride and the other through the buillder.apply(name,...)?

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java [hunk 11]
```diff
@@ -26,55 +26,89 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.regex.Pattern;
 import javax.annotation.Nullable;
 import org.flyte.api.v1.LiteralType;
 import org.flyte.api.v1.SimpleType;
 import org.flyte.api.v1.WorkflowTemplate;
 
 public class SdkWorkflowBuilder {
+
+  private static final Pattern PATTERN = Pattern.compile("([a-z])([A-Z]+)");
```
Comment (ckiosidis) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java:42]: we allow for capitals? i thought that only lower case was allowed

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java (hunk 10 above)
Comment (honnix) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkTransform.java:76]: `buillder.apply(nodeId,...)` is for setting node ID, not node name.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java (hunk 11 above)
Comment (honnix) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java:42]: This is to convert `FooBar` to `foo-bar` for example.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java (hunk 8 above)
Comment (honnix) [reply flytekit-java/src/main/java/org/flyte/flytekit/SdkWorkflowBuilder.java]: OK, I revised it to use alphabet as prefix. This will also decrease the chance of conflicting.

**Review by ckiosidis — APPROVED** [review]
looks ok to me 👍

**Review by rafaelraposospot — COMMENTED** [review]
In flytekit-python, it throws exception if not set and there's a method to convert to dns-compliant.

See: 
https://github.com/flyteorg/flytekit/blob/master/flytekit/core/node.py#L28
https://github.com/flyteorg/flytekit/blob/master/flytekit/core/utils.py#L13

Should we align with the Python way? Fine either way for me just want to be sure if we're supposed to be as much aligned as possible or if we can have our own flavor.

**Comment by honnix** [comment]
Where should we throw? In case of an invalid node id? That is reasonable but seems to be better in a different PR.

File: flytekit-java/src/main/java/org/flyte/flytekit/SdkPartialTransform.java [hunk 12]
```diff
@@ -94,10 +94,22 @@ public SdkTransform withUpstreamNode(SdkNode node) {
...
 
     SdkNodeMetadata newMetadata = SdkNodeMetadata.builder().name(name).build();
-    checkForDuplicateMetadata(metadata, newMetadata, SdkNodeMetadata::name, "name");
+    try {
+      checkForDuplicateMetadata(metadata, newMetadata, SdkNodeMetadata::name, "name");
+    } catch (IllegalArgumentException e) {
+      if (failOnDuplicate) {
+        throw e;
+      }
+      return this;
```
Comment (narape) [inline flytekit-java/src/main/java/org/flyte/flytekit/SdkPartialTransform.java:111]: This is surprising for me, if a name is already set and check is false then we don't override the name (this method becomes no op).

My suggestion would be to create a method called `withNameOverrideIfNotSet` (or similar) and then write the method like this:
```java
  public SdkTransform withNameOverrideIfNotSet(String name) {
    if (metadata.name() != null) {
      return this;
    }
    return withNameOverride(name);
  }
```
Leaving the `withNameOverride` as it was original

**Comment by honnix** [comment]
Please feel free to take over and make necessary changes. I will not have time to work on this any time soon. Thank you.

**Review by ckiosidis — APPROVED** [review]
looks good
